### ISSUE TITLE
Error finetuning with Japanese voice setting.

### ISSUE THREAD

**Issue by DeathmonkeyJ** [erew123/alltalk_tts#105]
[diagnostics.log](https://github.com/erew123/alltalk_tts/files/14415248/diagnostics.log)

> **Getting this error while finetuning using Japanese voice setting, not well versed in this stuff so I'm lost. I'm using a venv for this**

Traceback (most recent call last):
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 998, in preprocess_dataset
    train_meta, eval_meta, audio_total_size = format_audio_list(target_language=language, whisper_model=whisper_model, out_path=out_path, eval_split_number=eval_split_number, gradio_progress=progress)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 229, in format_audio_list
    sentence = multilingual_cleaners(sentence, target_language)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\TTS\tts\layers\xtts\tokenizer.py", line 558, in multilingual_cleaners
    text = expand_numbers_multilingual(text, lang)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\TTS\tts\layers\xtts\tokenizer.py", line 538, in expand_numbers_multilingual
    text = re.sub(_ordinal_re[lang], lambda m: _expand_ordinal(m, lang), text)
                  ~~~~~~~~~~~^^^^^^
KeyError: 'ja'

> **After this, I ran it as an English dataset and got it to finish processing, after the initial loading I was able to load it as Japanese. But I began getting this error during training.**

  TRAINING (2024-02-27 02:18:29)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\spawn.py", line 122, in spawn_main
 ! Run is kept in C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune\tmp-trn\training\XTTS_FT-February-27-2024_02+16AM-0000000
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
EOFError: Ran out of input
Traceback (most recent call last):
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1833, in fit
    self._fit()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1785, in _fit
    self.train_epoch()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1503, in train_epoch
    for cur_step, batch in enumerate(self.train_loader):
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 439, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 387, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 1040, in __init__
    w.start()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\popen_spawn_win32.py", line 94, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "stringsource", line 2, in fugashi.fugashi.Tagger.__reduce_cython__
TypeError: self.c_tagger cannot be converted to a Python object for pickling
Traceback (most recent call last):
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1833, in fit
    self._fit()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1785, in _fit
    self.train_epoch()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1503, in train_epoch
    for cur_step, batch in enumerate(self.train_loader):
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 439, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 387, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\torch\utils\data\dataloader.py", line 1040, in __init__
    w.start()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\popen_spawn_win32.py", line 94, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "stringsource", line 2, in fugashi.fugashi.Tagger.__reduce_cython__
TypeError: self.c_tagger cannot be converted to a Python object for pickling

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 1102, in train_model
    config_path, original_xtts_checkpoint, vocab_file, exp_path, speaker_wav = train_gpt(language, num_epochs, batch_size, grad_acumm, train_csv, eval_csv, output_path=str(output_path), max_audio_length=max_audio_length)
                                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 444, in train_gpt
    trainer.fit()
  File "C:\Users\Jace\miniconda3\envs\alltalkenv\Lib\site-packages\trainer\trainer.py", line 1862, in fit
    sys.exit(1)
SystemExit: 1

**Comment by erew123** [comment]
Hi @DeathmonkeyJ 

I've done my best to research this for you and I'm guessing your system is not a native Japanese Windows system.

The issues are occurring within Python scripts that are outside of AllTalk and therefore what I maintain/have access to, so that makes it a bit more complicated of course. However, I have done my best to look through those and check them. 

First off your system does have the base requirements for working with Japanese, as far as presenting the base character set on screen and general TTS generation. These are `cutlett` and `unidic-lite`. I've double confirmed the Whisper model (v2 and v3) used in Step 1 does have Japanese support (at least as best I can tell from their research papers its all in the 1x model). https://cdn.openai.com/papers/whisper.pdf (pages 23 to 27)

However, there is one suggested package I have not set to install with AllTalk, which is mecab https://github.com/coqui-ai/TTS/blob/dev/requirements.ja.txt and the initial error seems to suggest that input support for Japanese is not on your system and some kind of language localisation issue, so mecab may well be required on a non-Japanese Windows system. https://pypi.org/project/mecab-python3/

As for the second issue, reading further into some of these scripts, it **may** call on fugashi https://github.com/polm/fugashi?tab=readme-ov-file#installing-a-dictionary though I am slightly speculating on this.

So my best effort suggestion here is to drop to a command prompt and inside the AllTalk folder run the `start_environment.bat` file which should start the python environment and you should see that represented on the prompt e.g. (it would look similar to this, though this is for text-generation-webui, so I'm just using it as a representation so you can see how a command prompt would change).

![image](https://github.com/erew123/alltalk_tts/assets/35898566/e6008a29-2c0f-43e3-bd7c-56c5ea4387bb)

Once the environment is started, I would try the following commands (one at a time):

`pip install mecab-python3`

And I assume you may need the following too *BUT* you may want to try the Step 1 of finetuning after the mecab install and see if that resolves the problem.

`pip install 'fugashi[unidic]'`

`python -m unidic download`

(please note this says there will be a 700MB download with `unidic download`). 

All should assumedly go through and install ok.

I cannot discount this being an issue/error on those external scripts in how they are handling Japanese, but with no access to change them, I'm a bit limited there as to how far I can go with this.

I guess give that a go and see where you get to, though, I may be limited as to what suggestions we can try from here on if that doesn't solve it.

Thanks

**Comment by DeathmonkeyJ** [comment]
> Thanks for looking into this for me btw, much appreciated. After going through those steps, part one still had the same errors, and using English setting to then proceed to the second step it had more errors than before somehow I think?

Traceback (most recent call last):
  File "<string>", line 1, in <module>
 ! Run is kept in C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune\tmp-trn\training\XTTS_FT-February-28-2024_07+48PM-0000000
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
EOFError: Ran out of input
Traceback (most recent call last):
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1833, in fit
    self._fit()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1785, in _fit
    self.train_epoch()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1503, in train_epoch
    for cur_step, batch in enumerate(self.train_loader):
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 438, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 386, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 1039, in __init__
    w.start()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\popen_spawn_win32.py", line 95, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "stringsource", line 2, in fugashi.fugashi.Tagger.__reduce_cython__
TypeError: self.c_tagger cannot be converted to a Python object for pickling
Traceback (most recent call last):
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1833, in fit
    self._fit()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1785, in _fit
    self.train_epoch()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1503, in train_epoch
    for cur_step, batch in enumerate(self.train_loader):
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 438, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 386, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\torch\utils\data\dataloader.py", line 1039, in __init__
    w.start()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\context.py", line 336, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\popen_spawn_win32.py", line 95, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
  File "stringsource", line 2, in fugashi.fugashi.Tagger.__reduce_cython__
TypeError: self.c_tagger cannot be converted to a Python object for pickling

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 1102, in train_model
    config_path, original_xtts_checkpoint, vocab_file, exp_path, speaker_wav = train_gpt(language, num_epochs, batch_size, grad_acumm, train_csv, eval_csv, output_path=str(output_path), max_audio_length=max_audio_length)
                                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\finetune.py", line 444, in train_gpt
    trainer.fit()
  File "C:\Users\Jace\Desktop\alltalk_tts-main\alltalk_tts-main\alltalk_environment\env\Lib\site-packages\trainer\trainer.py", line 1862, in fit
    sys.exit(1)
SystemExit: 1

**Comment by erew123** [comment]
Hi @DeathmonkeyJ 

I think I have a possible solution. It looks as though the Japanese language packs that we have now installed are doing their job (on Step 2) however, they cannot handle multiprocessing (so it appears). This is mixed issue between the fugashi Python library and Coqui's scripts, neither of which I have direct control over.

However, we can attempt to disable Step 2 using multiprocessing. It will be a bit slower by ? amount, but it should work, in theory.

I don't have a Japanese training set here and as mentioned, not knowing Japanese I would struggle to know if its splitting things down and training correctly. But I can either give you something to try OR I can try building a updated finetune if you prefer.

My suggestion is to edit line 399 of finetune.py in a text editor. and change `num_loader_workers=8,` to `num_loader_workers=0,` 

So basically changing the 8 to a 0 

![image](https://github.com/erew123/alltalk_tts/assets/35898566/71ac9360-9a5e-433e-b869-3a7fec7f8a4c)

This should force their scripts to not use multiprocessing..... in theory!

I believe it will be a 0 and not a 1, so you may want to try the 0 first and see if that works.

If you don't feel comfortable changing the setting in the file yourself, please let me know and I will upload a changed copy somewhere for you to download. 

If you do try it, if you would let me know if it did or didn't work, then I should be able to put something more permanent in place.

Thanks

**Comment by erew123** [comment]
Hi @DeathmonkeyJ 

Is this still an issue for you? Have you managed to attempt the change I suggested?

Thanks

**Comment by DeathmonkeyJ** [comment]
Sorry about the delayed response, this did fix the error! Now I'm just running out of memory on my card lol. On a side note I'm getting this error during generation but can't tell if it is actually causing issues. 

[!] Warning: The text length exceeds the character limit of 71 for language 'ja', this might cause truncated audio

**Comment by erew123** [comment]
Thanks for reply! Good to know that worked, I should be able to implement a permanent fix on that issue of training Japanese.

For the separate warning message, thats a limit/warning message set by Coqui and not myself. 71 is a lot smaller than I expected (its 250 characters in English/Latin alphabet and apparently 213 in Italian). My understanding though is these are limitations of the model. I assume because the Japanese character set is a larger alphabet and the way characters are represented (within a computer) uses more data per character (I believe), hence the 250 drops down to a smaller number.

Now, saying that, these limitations are not a case of the model stopping working at that point. If you go past these amount of characters it will, in English at least, keep generating the text to speech, but the voice is more susceptible to deviate from the original voice sample and may produce strange sounding reproduction. So if its sounding ok, then you shouldn't have an issue. Though there could be benefit in splitting down the text into smaller chunks, but that would require someone who can code and work with the Japanese language to come up with code that can split it best. I have done this to a degree for english/latin but I cant for Japanese, because Id need to learn Japanese to understand if its doing the correct splitting.

At least the training is working for you. Im going to close this ticket for now, but feel free to reply again if you need to.

Thanks

**Comment by erew123** [comment]
Hi @DeathmonkeyJ 

FYI, I have updated finetuning and also added a custom tokenizer to handle Japanese (hopefully properly).

Thanks
